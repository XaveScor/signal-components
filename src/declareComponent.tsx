import React from "react";
import { Atom, AtomState, Ctx } from "@reatom/core";
import { reatomComponent, useCtx } from "@reatom/npm-react";
import { InsideProps, OutsideProps } from "./types";
import { createPropsProxy } from "./innerProps";
import { createComponentsStore } from "./componentsStore";
import { Arg, createWireHook } from "./wireHook";
import { createBindCtx } from "./bindCtx";

export type ReturnComponent<Props> = React.FC<OutsideProps<Props>>;
type ComponentF = {
  <
    A extends Atom<string | number | boolean | null | undefined>,
    B extends string | number | boolean | null | undefined,
  >(
    anAtom: A,
    mapper: (value: AtomState<A>) => B,
  ): React.ReactElement;
  <A extends Atom<string | number | boolean | null | undefined>>(
    anAtom: A,
  ): React.ReactElement;
};
type RenderArg = {
  reatomCtx: Ctx;
  spy: Arg["spy"];
  component: ComponentF;
};
type RenderF = (arg: RenderArg) => React.ReactElement;
type ComponentArg<Props> = {};
type ComponentReturnType = RenderF | React.JSX.Element;
type Component<Props> = (
  props: InsideProps<Props>,
  arg: ComponentArg<Props>,
) => ComponentReturnType;

export function declareComponent<Props>(
  component: Component<Props>,
): ReturnComponent<Props> {
  // @ts-ignore TODO: implement ref support later
  return (props: OutsideProps<Props>) => {
    const rootCtx = useCtx();
    const { insideProps, setProps } = React.useMemo(
      () => createPropsProxy<Props>(rootCtx, props),
      [],
    );
    setProps(props);
    const [, setState] = React.useState(0);
    const rerender = React.useCallback(() => setState((s) => s + 1), []);

    const { rewireHooks, render, runWires } = React.useMemo(
      () => createWireHook({ rerender, ctx: rootCtx }),
      [],
    );
    const { runBinded } = React.useMemo(
      () => createBindCtx({ ctx: rootCtx }),
      [],
    );
    const firstRender = React.useRef<ComponentReturnType | null>(null);
    if (!firstRender.current) {
      runBinded(() => {
        runWires(() => {
          firstRender.current = component(insideProps, {});
        });
      });
    } else {
      render();
    }
    React.useEffect(() => rewireHooks(), []);
    // because runWires is sync
    const wrapped = firstRender.current!;

    if (typeof wrapped !== "function") {
      return wrapped;
    }

    const InitPhase = React.useMemo(
      () =>
        React.memo(() => {
          const componentsStore = createComponentsStore();

          const RenderPhase = reatomComponent(({ ctx }) => {
            return wrapped({
              reatomCtx: ctx,
              component: componentsStore.renderAtom,
              spy: (anAtom) => ctx.spy(anAtom),
            });
          });

          return <RenderPhase />;
        }),
      [],
    );

    return <InitPhase />;
  };
}
