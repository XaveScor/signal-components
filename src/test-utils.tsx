import React, { StrictMode } from "react";
import { createTestCtx, TestCtx } from "@reatom/testing";
import { render } from "@testing-library/react";
import { reatomContext } from "@reatom/npm-react";

export function customRender(element: React.ReactElement) {
  const ctx = createTestCtx();
  const rendered = render(element, {
    wrapper: ({ children }) => (
      <StrictMode>
        <reatomContext.Provider value={ctx}>{children}</reatomContext.Provider>
      </StrictMode>
    ),
  });

  return { rendered, ctx };
}
