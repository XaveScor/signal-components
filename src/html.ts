import React from "react";
import { ReturnComponent } from "./declareComponent";
import { wrap } from "./wrap";

const cache = new Map<
  keyof React.JSX.IntrinsicElements,
  ReturnComponent<any>
>();

export const html = new Proxy<{
  [K in keyof React.JSX.IntrinsicElements]: ReturnComponent<
    React.JSX.IntrinsicElements[K]
  >;
}>(
  // @ts-ignore
  {},
  {
    get(target, tag: keyof React.JSX.IntrinsicElements) {
      if (cache.has(tag)) {
        return cache.get(tag);
      }
      const ret = wrap(tag);
      cache.set(tag, ret);
      return ret;
    },
  },
);
