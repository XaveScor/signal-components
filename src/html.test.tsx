import React, {
  ButtonHTMLAttributes,
  ComponentProps,
  JSX,
  MouseEventHandler,
  SyntheticEvent,
} from "react";
import { describe, test, expect, expectTypeOf } from "vitest";
import { html } from "./html";
import { customRender } from "./test-utils";
import { act } from "@testing-library/react";
import { atom } from "@reatom/core";
import { ReturnComponent } from "./declareComponent";

describe("html", () => {
  test("span", () => {
    const x = atom(1);
    const { ctx, rendered } = customRender(<html.span>{x}</html.span>);
    expect(rendered.asFragment()).toMatchSnapshot();

    act(() => {
      x(ctx, 2);
    });
    expect(rendered.asFragment()).toMatchSnapshot();

    rendered.rerender(<html.span>10</html.span>);
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  test("div", () => {
    const x = atom(1);
    const { ctx, rendered } = customRender(<html.div>{x}</html.div>);
    expect(rendered.asFragment()).toMatchSnapshot();

    act(() => {
      x(ctx, 2);
    });
    expect(rendered.asFragment()).toMatchSnapshot();

    rendered.rerender(<html.div>10</html.div>);
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  describe("#34 bug", () => {
    test("onClick prop should infer right type", () => {
      type ButtonProps = ComponentProps<typeof html.button>;

      expectTypeOf<NonNullable<ButtonProps["onClick"]>>().toMatchTypeOf<
        (e: React.MouseEvent<HTMLButtonElement>) => void
      >();
    });

    test("ReturnComponent should infer right on* type", () => {
      type Component = ReturnComponent<{ onClick: (x: number) => void }>;
      type Props = ComponentProps<Component>;

      expectTypeOf<NonNullable<Props["onClick"]>>().toMatchTypeOf<
        (x: number) => void
      >();
    });

    test("ReturnComponent should infer right on* type for JSX.IntrinsicElements", () => {
      type Component = ReturnComponent<JSX.IntrinsicElements["button"]>;
      type Props = ComponentProps<Component>;

      expectTypeOf<NonNullable<Props["onClick"]>>().toMatchTypeOf<
        (e: React.MouseEvent<HTMLButtonElement>) => void
      >();
    });

    test("ReturnComponent should infer right on* type for JSX.IntrinsicElements", () => {
      type Component = ReturnComponent<JSX.IntrinsicElements["button"]>;
      type Props = ComponentProps<Component>;

      expectTypeOf<NonNullable<Props["onClick"]>>().toMatchTypeOf<
        (e: React.MouseEvent<HTMLButtonElement>) => void
      >();
    });

    test("ReturnComponent should infer right on* type for JSX.IntrinsicElements['onclick']", () => {
      type Component = ReturnComponent<{
        onClick: JSX.IntrinsicElements["button"]["onClick"];
      }>;
      type Props = ComponentProps<Component>;

      expectTypeOf<NonNullable<Props["onClick"]>>().toMatchTypeOf<
        (e: React.MouseEvent<HTMLButtonElement>) => void
      >();
    });

    test("ReturnComponent should infer right on* type for JSX.IntrinsicElements['onclick']", () => {
      type Component = ReturnComponent<{
        onClick: JSX.IntrinsicElements["button"]["onClick"];
      }>;
      type Props = ComponentProps<Component>;

      expectTypeOf<NonNullable<Props["onClick"]>>().toMatchTypeOf<
        (e: React.MouseEvent<HTMLButtonElement>) => void
      >();
    });

    test("ReturnComponent should infer right on* type for ButtonHTMLAttributes", () => {
      type Component = ReturnComponent<{
        onClick: ButtonHTMLAttributes<HTMLButtonElement>["onClick"];
      }>;
      type Props = ComponentProps<Component>;

      expectTypeOf<NonNullable<Props["onClick"]>>().toMatchTypeOf<
        (e: React.MouseEvent<HTMLButtonElement>) => void
      >();
    });

    test("ReturnComponent should infer right on* type for MouseEventHandler", () => {
      type Component = ReturnComponent<{
        onClick: MouseEventHandler<HTMLButtonElement> | undefined;
      }>;
      type Props = ComponentProps<Component>;

      expectTypeOf<NonNullable<Props["onClick"]>>().toMatchTypeOf<
        (e: React.MouseEvent<HTMLButtonElement>) => void
      >();
    });

    test("ReturnComponent should infer right on* type for bivarianceHack", () => {
      type EventHandler<E extends SyntheticEvent<any>> = {
        bivarianceHack(event: E): void;
      }["bivarianceHack"];
      type Component = ReturnComponent<{
        onClick: EventHandler<React.MouseEvent<HTMLButtonElement>> | undefined;
      }>;
      type Props = ComponentProps<Component>;

      expectTypeOf<NonNullable<Props["onClick"]>>().toMatchTypeOf<
        ((e: React.MouseEvent<HTMLButtonElement>) => void) | undefined
      >();
    });

    test("ReturnComponent should infer right on* type for MouseEvent", () => {
      type Component = ReturnComponent<{
        onClick: (e: React.MouseEvent<HTMLButtonElement>) => void;
      }>;
      type Props = ComponentProps<Component>;

      expectTypeOf<Props["onClick"]>().toMatchTypeOf<
        (e: React.MouseEvent<HTMLButtonElement>) => void
      >();
    });
  });
});
