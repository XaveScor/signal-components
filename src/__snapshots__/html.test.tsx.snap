// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`html > div 1`] = `
<DocumentFragment>
  <div>
    1
  </div>
</DocumentFragment>
`;

exports[`html > div 2`] = `
<DocumentFragment>
  <div>
    2
  </div>
</DocumentFragment>
`;

exports[`html > div 3`] = `
<DocumentFragment>
  <div>
    10
  </div>
</DocumentFragment>
`;

exports[`html > span 1`] = `
<DocumentFragment>
  <span>
    1
  </span>
</DocumentFragment>
`;

exports[`html > span 2`] = `
<DocumentFragment>
  <span>
    2
  </span>
</DocumentFragment>
`;

exports[`html > span 3`] = `
<DocumentFragment>
  <span>
    10
  </span>
</DocumentFragment>
`;
